File: bench/poly/poly.rs
use std::time::Instant;

const N: usize = 8192;
const N_ITERS: usize = 80;
const N_RUNS: usize = 21;
const N_WARMUP: usize = 5;

fn mix(h: u32, x: u32) -> u32 {
    (h ^ x).wrapping_mul(0x0100_0193)
}

fn median_us(samples: &mut [f64]) -> u64 {
    for i in 1..samples.len() {
        let v = samples[i];
        let mut j = i;
        while j > 0 && samples[j - 1] > v {
            samples[j] = samples[j - 1];
            j -= 1;
        }
        samples[j] = v;
    }
    (samples[(samples.len() - 1) >> 1] * 1000.0) as u64
}

fn init(f64: &mut [f64], i32: &mut [i32]) {
    for i in 0..N {
        f64[i] = (i % 251) as f64 * 0.25;
        i32[i] = ((i * 17) & 1023) as i32;
    }
}

fn sum_f64(xs: &[f64]) -> f64 {
    let mut s = 0.0;
    for &x in xs { s += x; }
    s
}

fn sum_i32(xs: &[i32]) -> i32 {
    let mut s = 0i32;
    for &x in xs { s += x; }
    s
}

fn run_kernel(f64: &[f64], i32: &[i32]) -> u32 {
    let mut h = 0x811c_9dc5u32;
    for _ in 0..N_ITERS {
        h = mix(h, sum_f64(f64) as u32);
        h = mix(h, sum_i32(i32) as u32);
    }
    h
}

fn main() {
    let mut f64 = vec![0.0f64; N];
    let mut i32 = vec![0i32; N];
    init(&mut f64, &mut i32);

    let mut cs = 0u32;
    for _ in 0..N_WARMUP { cs = run_kernel(&f64, &i32); }

    let mut samples = [0.0; N_RUNS];
    for sample in &mut samples {
        let t0 = Instant::now();
        cs = run_kernel(&f64, &i32);
        *sample = t0.elapsed().as_secs_f64() * 1000.0;
    }

    println!(
        "median_us={} checksum={} samples={} stages={} runs={}",
        median_us(&mut samples), cs, N * N_ITERS * 2, 2, N_RUNS
    );
}

File: bench/poly/poly.zig
const std = @import("std");
const Io = std.Io;

const N = 8192;
const N_ITERS = 80;
const N_RUNS = 21;
const N_WARMUP = 5;

fn nowMs() f64 {
    var ts: std.c.timespec = undefined;
    _ = std.c.clock_gettime(std.c.CLOCK.MONOTONIC, &ts);
    return @as(f64, @floatFromInt(ts.sec)) * 1000.0 + @as(f64, @floatFromInt(ts.nsec)) / 1_000_000.0;
}

fn mix(h: u32, x: u32) u32 {
    return (h ^ x) *% 0x01000193;
}

fn medianUs(samples: *[N_RUNS]f64) u64 {
    var i: usize = 1;
    while (i < samples.len) : (i += 1) {
        const v = samples[i];
        var j = i;
        while (j > 0 and samples[j - 1] > v) : (j -= 1) samples[j] = samples[j - 1];
        samples[j] = v;
    }
    return @as(u64, @intFromFloat(samples[(samples.len - 1) >> 1] * 1000.0));
}

fn init(f64s: []f64, i32s: []i32) void {
    var i: usize = 0;
    while (i < N) : (i += 1) {
        f64s[i] = @as(f64, @floatFromInt(i % 251)) * 0.25;
        i32s[i] = @as(i32, @intCast((i *% 17) & 1023));
    }
}

fn sumF64(xs: []const f64) f64 {
    var s: f64 = 0;
    for (xs) |x| s += x;
    return s;
}

fn sumI32(xs: []const i32) i32 {
    var s: i32 = 0;
    for (xs) |x| s +%= x;
    return s;
}

fn runKernel(f64s: []const f64, i32s: []const i32) u32 {
    var h: u32 = 0x811c9dc5;
    var i: usize = 0;
    while (i < N_ITERS) : (i += 1) {
        h = mix(h, @as(u32, @bitCast(@as(i32, @intFromFloat(sumF64(f64s))))));
        h = mix(h, @as(u32, @bitCast(sumI32(i32s))));
    }
    return h;
}

pub fn main(init_args: std.process.Init) !void {
    const io = init_args.io;
    var stdout_buffer: [256]u8 = undefined;
    var stdout_writer = Io.File.stdout().writer(io, &stdout_buffer);
    const stdout = &stdout_writer.interface;

    const allocator = std.heap.page_allocator;
    const f64s = try allocator.alloc(f64, N);
    const i32s = try allocator.alloc(i32, N);
    defer allocator.free(f64s);
    defer allocator.free(i32s);
    init(f64s, i32s);

    var cs: u32 = 0;
    var i: usize = 0;
    while (i < N_WARMUP) : (i += 1) cs = runKernel(f64s, i32s);

    var samples = [_]f64{0} ** N_RUNS;
    i = 0;
    while (i < N_RUNS) : (i += 1) {
        const t0 = nowMs();
        cs = runKernel(f64s, i32s);
        samples[i] = nowMs() - t0;
    }
    try stdout.print("median_us={d} checksum={d} samples={d} stages={d} runs={d}\n", .{ medianUs(&samples), cs, N * N_ITERS * 2, 2, N_RUNS });
    try stdout.flush();
}
